File: 266_Palindrome_Permutation.java
/*
https://leetcode.com/problems/palindrome-permutation/
Given a string, determine if a permutation of the string could form a palindrome.

Add each character to a set.  If caharcter is seen twice then remove it from set.  Set contains all chars with and
odd count.  Palindrome can have at most one such char.
Time - O(n)
Space - O(n)
*/

public class Solution {
    public boolean canPermutePalindrome(String s) {

        Set<Character> charFound = new HashSet<Character>();

        for (int i = 0; i < s.length(); ++i) {

            char c = s.charAt(i);
            if (charFound.contains(c))
                charFound.remove(c);
            else
                charFound.add(c);
        }

        return charFound.size() <= 1;
    }
}

File: python_1_to_1000/266_Palindrome_Permutation.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/palindrome-permutation/
# Given a string, determine if a permutation of the string could form a palindrome.

# There can be at most one character with an odd count in a palindrome.
# Time - O(n)
# Space - O(1)

from collections import Counter

class Solution(object):
    def canPermutePalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        freq = Counter(s)
        odd = False                 # no odd has been seen

        for letter, count in freq.items():
            if count % 2 == 1:
                if odd:             # amother char has odd count
                    return False
                odd = True

        return True